File: templates/Virtual-Assistant-Template/typescript/samples/sample-skill/src/adapters/customSkillAdapter.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

import {
    AutoSaveStateMiddleware,
    BotTelemetryClient,
    ConversationState,
    StatePropertyAccessor,
    TelemetryLoggerMiddleware,
    TranscriptLoggerMiddleware,
    TranscriptStore,
    UserState} from 'botbuilder';
import { AzureBlobTranscriptStore } from 'botbuilder-azure';
import { DialogState } from 'botbuilder-dialogs';
import {
    // PENDING: SkillHttpBotAdapter should be replaced with SkillWebSocketBotAdapter
    SkillHttpBotAdapter,
    SkillMiddleware } from 'botbuilder-skills';
import { EventDebuggerMiddleware, SetLocaleMiddleware } from 'botbuilder-solutions';
import { IBotSettings } from '../services/botSettings';

// PENDING: this class should extends from SkillWebSocketBotAdapter instead of SkillHttpBotAdapter
export class CustomSkillAdapter extends SkillHttpBotAdapter {

    public constructor(
        settings: Partial<IBotSettings>,
        userState: UserState,
        conversationState: ConversationState,
        telemetryClient: BotTelemetryClient,
        dialogStateAccessor: StatePropertyAccessor<DialogState>
    ) {
        super(telemetryClient);
        // PENDING: the next line should be uncommented when the WS library is merged
        // super();
        if (settings.blobStorage === undefined) {
            throw new Error('There is no blobStorage value in appsettings file');
        }

        const transcriptStore: TranscriptStore = new AzureBlobTranscriptStore({
            containerName: settings.blobStorage.container,
            storageAccountOrConnectionString: settings.blobStorage.connectionString
        });

        // Uncomment the following line for local development without Azure Storage
        // this.use(new TranscriptLoggerMiddleware(new MemoryTranscriptStore()));
        this.use(new TelemetryLoggerMiddleware(telemetryClient, true));
        this.use(new TranscriptLoggerMiddleware(transcriptStore));
        this.use(new SetLocaleMiddleware(settings.defaultLocale || 'en-us'));
        this.use(new EventDebuggerMiddleware());
        this.use(new AutoSaveStateMiddleware(conversationState, userState));
        this.use(new SkillMiddleware(conversationState, dialogStateAccessor));
    }
}

File: samples/assistants/AdaptiveAssistant/AdaptiveCalendarSkill/AdaptiveCalendarSkill/Adapters/CustomSkillAdapter.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using AdaptiveCalendarSkill.Services;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Dialogs.Declarative.Resources;
using Microsoft.Bot.Builder.Dialogs.Declarative.Types;
using Microsoft.Bot.Builder.LanguageGeneration;
using Microsoft.Bot.Builder.Skills;
using Microsoft.Bot.Builder.Solutions.Middleware;
using Microsoft.Extensions.Configuration;

namespace AdaptiveCalendarSkill.Adapters
{
    public class CustomSkillAdapter : SkillWebSocketBotAdapter
    {
        public CustomSkillAdapter(
            IConfiguration configuration,
            BotSettings settings,
            IStorage storage,
            UserState userState,
            ConversationState conversationState,
            ResourceExplorer resourceExplorer,
            IBotTelemetryClient telemetryClient)
        {
            OnTurnError = async (turnContext, exception) =>
            {
                await turnContext.SendActivityAsync($"{exception.Message}");
                await turnContext.SendActivityAsync($"{exception.StackTrace}");
                await turnContext.SendActivityAsync("Sorry, something went wrong (skill adapter)!");
            };

            TypeFactory.Configuration = configuration;
            this.UseStorage(storage);
            this.UseState(userState, conversationState);
            this.UseLanguageGeneration(resourceExplorer);
            // this.UseDebugger(configuration.GetValue("debugport", 4712), events: new Events<AdaptiveEvents>());

            // Uncomment the following line for local development without Azure Storage
            // Use(new TranscriptLoggerMiddleware(new MemoryTranscriptStore()));
            Use(new TelemetryLoggerMiddleware(telemetryClient, logPersonalInformation: true));
            Use(new SetLocaleMiddleware(settings.DefaultLocale ?? "en-us"));
            Use(new EventDebuggerMiddleware());
            Use(new SkillMiddleware(userState, conversationState, conversationState.CreateProperty<DialogState>(nameof(AdaptiveCalendarSkill))));
        }
    }
}
